internal/xds/bootstrap/tlscreds/bundle_ext_test.go

```diff
@@ -224,6 +229,87 @@ func (s) TestCaReloading(t *testing.T) {
 	}
 }
 
+func (s) TestSPIFFEReloading(t *testing.T) {
+	clientSPIFFEBundle, err := os.ReadFile(testdata.Path("spiffe_end2end/client_spiffebundle.json"))
+	if err != nil {
+		t.Fatalf("Failed to read test CA cert: %s", err)
+	}
+
+	// Write CA certs to a temporary file so that we can modify it later.
+	spiffePath := t.TempDir() + "/client_spiffe.json"
+	if err = os.WriteFile(spiffePath, clientSPIFFEBundle, 0644); err != nil {
+		t.Fatalf("Failed to write test SPIFFE Bundle %v: %v", clientSPIFFEBundle, err)
+	}
+	cfg := fmt.Sprintf(`{
+		"spiffe_trust_bundle_map_file": "%s",
+		"refresh_interval": ".01s"
+	}`, spiffePath)
+	tlsBundle, stop, err := tlscreds.NewBundle([]byte(cfg))
+	if err != nil {
+		t.Fatalf("Failed to create TLS bundle: %v", err)
+	}
+	defer stop()
+
+	serverCredentials := grpc.Creds(testutils.CreateServerTLSCredentialsCompatibleWithSPIFFE(t, tls.NoClientCert))
+	server := stubserver.StartTestService(t, nil, serverCredentials)
+	defer server.Stop()
+
+	conn, err := grpc.NewClient(
+		server.Address,
+		grpc.WithCredentialsBundle(tlsBundle),
+		grpc.WithAuthority("x.test.example.com"),
+	)
+	if err != nil {
+		t.Fatalf("Error dialing: %v", err)
+	}
+	defer conn.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	client := testgrpc.NewTestServiceClient(conn)
+	if _, err = client.EmptyCall(ctx, &testpb.Empty{}); err != nil {
+		t.Errorf("Error calling EmptyCall: %v", err)
+	}
+	// close the server and create a new one to force client to do a new
+	// handshake.
+	server.Stop()
+
+	wrongBundle, err := os.ReadFile(testdata.Path("spiffe_end2end/server_spiffebundle.json"))
+	if err != nil {
+		t.Fatalf("Failed to read test spiffe bundle %v: %v", "spiffe_end2end/server_spiffebundle.json", err)
+	}
+	// unload root cert
+	err = os.WriteFile(spiffePath, wrongBundle, 0644)
+	if err != nil {
+		t.Fatalf("Failed to write test spiffe bundle %v: %v", "spiffe_end2end/server_spiffebundle.json", err)
+	}
+
+	for ; ctx.Err() == nil; <-time.After(10 * time.Millisecond) {
+		ss := stubserver.StubServer{
+			Address:    server.Address,
+			EmptyCallF: func(context.Context, *testpb.Empty) (*testpb.Empty, error) { return &testpb.Empty{}, nil },
+		}
+		server = stubserver.StartTestService(t, &ss, serverCredentials)
```

Comment:
Are you trying to stop a server and restart it with the same port. This can be flaky and it is possible that you are never able to get the same port again. Instead use the restartable listener that we have in our testutils. See: https://github.com/grpc/grpc-go/blob/master/internal/testutils/restartable_listener.go

Reply:
The restartable listener type provides a `Stop` and `Restart` method. When `Stop` is called, existing connections are closed and new connections are not accepted, and when `Restart` is called, new connections are accepted. Therefore, once `Restart` is called, we should see another handshake.

Reply:
Ah okay, I believe I've done it, thanks!